lab/cached_revision.py

```diff
@@ -0,0 +1,236 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import hashlib
+import logging
+import os.path
+import shutil
+import subprocess
+import tarfile
+
+from lab import tools
+
+
+GIT = "git"
+MERCURIAL = "hg"
+VERSION_CONTROL_SYSTEMS = [GIT, MERCURIAL]
+
+
+_ID_CACHE = {}
+
+
+def _get_id(cmd):
+    cmd = tuple(cmd)
+    if cmd not in _ID_CACHE:
+        try:
+            result = tools.get_string(subprocess.check_output(cmd).strip())
+        except subprocess.CalledProcessError:
+            logging.critical(
+                'Call failed: "{}". Please check path and revision.'.format(
+                    " ".join(cmd)
+                )
+            )
+        else:
+            assert result
+            _ID_CACHE[cmd] = result
+    return _ID_CACHE[cmd]
+
+
+def hg_id(repo, args=None, rev=None):
+    args = args or []
+    if rev:
+        args.extend(["-r", str(rev)])
+    cmd = ["hg", "id", "--repository", repo] + args
+    return _get_id(cmd)
+
+
+def git_id(repo, args=None, rev=None):
+    args = args or []
+    if rev:
+        args.append(rev)
+    cmd = [
+        "git",
+        "--git-dir",
+        os.path.join(repo, ".git"),
+        "rev-parse",
+        "--short",
+    ] + args
+    return _get_id(cmd)
+
+
+def _raise_unknown_vcs_error(vcs):
+    raise AssertionError('Unknown version control system "{}".'.format(vcs))
+
+
+def get_version_control_system(repo):
+    vcs = [
+        x
+        for x in VERSION_CONTROL_SYSTEMS
+        if os.path.exists(os.path.join(repo, ".{}".format(x)))
+    ]
+    if len(vcs) == 1:
+        return vcs[0]
+    else:
+        logging.critical(
+            "Repo {} must contain exactly one of the following subdirectories: {}".format(
+                repo, ", ".join(".{}".format(x) for x in VERSION_CONTROL_SYSTEMS)
+            )
+        )
+
+
+def get_global_rev(repo, rev=None):
+    vcs = get_version_control_system(repo)
+    if vcs == MERCURIAL:
+        return hg_id(repo, args=["-i"], rev=rev)
+    elif vcs == GIT:
+        return git_id(repo, rev=rev)
+    else:
+        _raise_unknown_vcs_error(vcs)
+
+
+def get_rev_id(repo, rev=None):
+    vcs = get_version_control_system(repo)
+    if vcs == MERCURIAL:
+        return hg_id(repo, rev=rev)
+    elif vcs == GIT:
+        return git_id(repo, rev=rev)
+    else:
+        _raise_unknown_vcs_error(vcs)
+
+
+def _compute_md5_hash(mylist):
+    m = hashlib.md5()
+    for s in mylist:
+        m.update(tools.get_bytes(s))
+    return m.hexdigest()[:8]
+
+
+class CachedRevision(object):
+    """This class represents checkouts of the used solver.
+
+    It provides methods for caching and compiling given revisions.
+    """
+
+    def __init__(self, repo, local_rev, build_script, build_options, exclude_dirs=[]):
```

Comment:
You should prob. use here exclude_dirs=None, and then on the first line of the init method just do `exclude_dirs = [] if exclude_dirs=None else exclude dirs`, or similar

Reply:
A more idiomatic solution would be `exclude_dirs = exclude_dirs or []`, but I haven't looked at the code.

Reply:
I just removed the default, it's not necessary